File: frontend-client/src/api/viewingTime.ts
import { apiFetch } from './client'

// -- Types --

export type EnforcementStatus = 'allowed' | 'warning_15' | 'warning_5' | 'blocked'
export type DeviceType = 'tv' | 'mobile' | 'tablet' | 'web'

export interface ViewingTimeBalance {
  profile_id: string
  is_child_profile: boolean
  has_limits: boolean
  used_minutes: number
  educational_minutes: number
  limit_minutes: number | null
  remaining_minutes: number | null
  is_unlimited_override: boolean
  next_reset_at: string | null
  warning_threshold_minutes: number[]
}

export interface HeartbeatRequest {
  profile_id: string
  session_id?: string
  title_id: string
  device_id: string
  device_type?: DeviceType
  is_paused?: boolean
}

export interface HeartbeatResponse {
  session_id: string
  enforcement: EnforcementStatus
  remaining_minutes: number | null
  used_minutes: number
  is_educational: boolean
}

export interface SessionEndResponse {
  session_id: string
  total_seconds: number
  ended_at: string
}

export interface PlaybackEligibility {
  eligible: boolean
  remaining_minutes: number | null
  reason: string | null
  next_reset_at: string | null
}

// -- API functions --

export function getBalance(profileId: string): Promise<ViewingTimeBalance> {
  return apiFetch(`/viewing-time/balance/${profileId}`)
}

export function sendHeartbeat(req: HeartbeatRequest): Promise<HeartbeatResponse> {
  return apiFetch('/viewing-time/heartbeat', {
    method: 'POST',
    body: JSON.stringify(req),
  })
}

export function endSession(sessionId: string): Promise<SessionEndResponse> {
  return apiFetch(`/viewing-time/session/${sessionId}/end`, {
    method: 'POST',
  })
}

export function checkPlaybackEligibility(profileId: string): Promise<PlaybackEligibility> {
  return apiFetch(`/viewing-time/playback-eligible/${profileId}`)
}

File: backend/app/schemas/viewing_time.py
"""Pydantic schemas for viewing time tracking endpoints."""

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class EnforcementStatus(str, Enum):
    allowed = "allowed"
    warning_15 = "warning_15"
    warning_5 = "warning_5"
    blocked = "blocked"


class DeviceType(str, Enum):
    tv = "tv"
    mobile = "mobile"
    tablet = "tablet"
    web = "web"


# -- Balance --

class ViewingTimeBalanceResponse(BaseModel):
    profile_id: uuid.UUID
    is_child_profile: bool
    has_limits: bool
    used_minutes: float
    educational_minutes: float
    limit_minutes: int | None
    remaining_minutes: float | None
    is_unlimited_override: bool
    next_reset_at: datetime | None
    warning_threshold_minutes: list[int] = Field(default_factory=lambda: [15, 5])


# -- Heartbeat --

class HeartbeatRequest(BaseModel):
    profile_id: uuid.UUID
    session_id: uuid.UUID | None = None
    title_id: uuid.UUID
    device_id: str = Field(max_length=100)
    device_type: DeviceType = DeviceType.web
    is_paused: bool = False


class HeartbeatResponse(BaseModel):
    session_id: uuid.UUID
    enforcement: EnforcementStatus
    remaining_minutes: float | None
    used_minutes: float
    is_educational: bool


# -- Session End --

class SessionEndResponse(BaseModel):
    session_id: uuid.UUID
    total_seconds: int
    ended_at: datetime


# -- Playback Eligibility --

class PlaybackEligibilityResponse(BaseModel):
    eligible: bool
    remaining_minutes: float | None = None
    reason: str | None = None
    next_reset_at: datetime | None = None
